photongo/Readme.md

```diff
@@ -122,356 +163,532 @@ enum Role {
+##### Update the role of an existing user
 
 ```go
-usr, err := client.User.Update(
-  user.New().Role(user.Role.ADMIN),
-  user.Where().ID("cjsyytzn0004d0982gbyeqep7"),
-)
+user, err := client.User.Update.Where(
+  photon.User.ID.Equals("cjsyytzn0004d0982gbyeqep7"),
+).Data(
+  photon.user.Role.Set(photon.User.Role.ADMIN),
+).Exec(ctx)
 ```
 
-#### Update the author of a post
+##### Update the author of a post
 
 ```go
-pst, err := client.Post.Update(
-  post.New().ConnectAuthor(
-    user.Connect().Email("bob@prisma.io"),
+post, err := client.Post.Update.Where(
+  photon.Post.ID.Equals("cjsx2j8bw02920b25rl806l07"),
+).Data(
+  photon.Post.Author.Connect.Where(
+    photon.User.Email.Equals("bob@prisma.io"),
   ),
-  post.Where().ID("cjsx2j8bw02920b25rl806l07"),
-)
+).Exec(ctx)
 ```
 
-### UpdateMany
+#### UpdateMany
 
-#### Update three posts by their IDs
+##### Update three posts by their IDs
 
 ```go
-updated, err := client.Post.UpdateMany(
-  post.New().Published(true),
-  post.Where().IDIn(
+updated, err := client.Post.UpdateMany.Where(
+  photon.Post.ID.In([]string{
     "cjsyqxwqv000l0982p5qdq34p",
     "cjsyqxwqo000j0982da8cvw7o",
     "cjsyqxwr0000n0982cke8i5sc",
-  ),
-)
+  }),
+).Data(
+  photon.Post.Published.Set(true),
+).Exec(ctx)
 ```
 
-#### Update all posts where the title contains the given string
+##### Update all posts where the title contains the given string
 
 ```go
-updated, err = client.Post.UpdateMany(
-  post.New().Published(true),
-  post.Where().TitleContains("prisma"),
-)
+updated, err = client.Post.UpdateMany.Where(
+  photon.Post.Title.Contains("prisma"),
+).Data(
+  photon.Post.Published.Set(true),
+).Exec(ctx)
 ```
 
-### Delete
+#### Delete
 
-#### Delete a post by its ID
+##### Delete a post by its ID
 
 ```go
-pst, err := client.Post.Delete(post.Where().ID(10))
+post, err := client.Post.Delete.Where(
+  photon.Post.ID.Equals(10),
+).Exec(ctx)
 ```
 
-#### Delete a user by their email
+##### Delete a user by their email
 
 ```go
-usr, err := client.User.Delete(user.Where().Email("alice@prisma.io"))
+user, err := client.User.Delete.Where(
+  photon.User.Email.Equals("alice@prisma.io"),
+).Exec(ctx)
 ```
 
-### DeleteMany
+#### DeleteMany
 
-#### Delete all posts that were created before 2018:
+##### Delete all posts that were created before 2018:
 
 ```go
-deleted, err := client.Post.DeleteMany(
-  post.Where().CreatedAtGt(christmas),
-)
+deleted, err := client.Post.DeleteMany.Where(
+  photon.Post.CreatedAt.Gt(christmas),
+).Exec(ctx)
 ```
 
-### Upsert
+#### Upsert
 
-#### Create a user or update their role
+##### Create a user or update their role
 
 ```go
-usr, err := client.User.Upsert(
-  user.New().Email("alice@prisma.io"),
-  user.New().Role(user.Role.ADMIN),
-  user.Where().Email("alice@prisma.io"),
-)
+user, err := client.User.Upsert.Where(
+  photon.User.Email.Equals("alice@prisma.io"),
+).Data(
+  photon.User.Email.Set("alice@prisma.io"),
+  photon.User.Role.Set(user.Role.ADMIN),
```

Comment:
If we go with the multi-package route, could we do `user.Role(...)`?

I thought about this last night and even though it's a **really** cool idea, I think it'll be too verbose in practice. We have to remember the context of where this code lives and who we're catering for. Even if we made photon incredibly type-safe in Go, the rest of their application logic won't be. Personally, I'd rather work with a succinct API and cater to professionals who write tests where these tests will catch the occasional runtime error.

Another tidbit: we can't ever make row-level violations type-safe (unique constraints). We're only tackling a subset of possible runtime errors with this API change.

Reply:
Sorry I should have clarified on what's too verbose: 

```go
// upsert the user by their slack ID
u := user.New().
	SlackID(slackUser.SlackID).
	Username(slackUser.Username).
	Email(slackUser.Email).
	FirstName(slackUser.FirstName).
	LastName(slackUser.LastName).
	AvatarURL(slackUser.AvatarURL).
	Timezone(slackUser.Timezone).
	IsTeamOwner(true).
	TeamID(tm.ID)
usr, err := user.UpsertBySlackIDAndTeamID(tx, res.UserID, tm.ID, u)
if err != nil {
	return errors.Wrap(err, "unable to upsert the user")
}
```

This is where I'm coming from, let's maybe reformat this in the new syntax and we can discuss further 🙏

Reply:
I'm pulling the dot syntax argument here again.. The only char difference you have is `.Set`, which is negligible. 

And, more importantly, my solution enforces the user to specify required fields with fixed parameters.

```go
// generated code
func (UserChain) Data(id UserIDParam, username UserUsernameParam /* ... */) {
  // ...
}
```

so you can do 

```go
user, err := client.User.Upsert.Where(
  photon.User.Email.Equals("alice@prisma.io"),
).Data(
  photon.User.ID.Set("alice@prisma.io"), // .ID returns UserIDParam, so you are forced to sepcify required fields
  photon.User.Email.Set("alice@prisma.io"),
  // ...
)
```

Reply:
I'd **love** to see examples of the exact syntactic differences. I think it's more than `.Set`. Here's my rough translation:

```go
u, err := user.UpsertBySlackIDAndTeamID(tx, res.UserID, tm.ID, user.New().
	SlackID(slackUser.SlackID).
	Username(slackUser.Username).
	Email(slackUser.Email).
	FirstName(slackUser.FirstName).
	LastName(slackUser.LastName).
	AvatarURL(slackUser.AvatarURL).
	Timezone(slackUser.Timezone).
	IsTeamOwner(true).
	TeamID(tm.ID),
)
```

vs.

```go
user, err := client.User.Upsert.Where(
  photon.User.SlackID.Equals("alice@prisma.io"),
  photon.User.TeamID.Equals("alice@prisma.io"),
).Data(
  photon.User.SlackID.Set(slackUser.SlackID),
	photon.User.Username.Set(slackUser.Username),
	photon.User.Email.Set(slackUser.Email),
	photon.User.FirstName.Set(slackUser.FirstName),
	photon.User.LastName.Set(slackUser.LastName),
	photon.User.AvatarURL.Set(slackUser.AvatarURL),
	photon.User.Timezone.Set(slackUser.Timezone),
	photon.User.IsTeamOwner.Set(true),
	photon.User.TeamID.Set(tm.ID),
)
```

Please correct if wrong 🙏

Reply:
^ disregard `UpsertBySlackIDAndTeamID `, it's more like `user.Where().SlackID().ID()`

Reply:
It's much better with dot imports:

```go
import (
  . "github.com/user/repo/photon"
)

// ...

user, err := client.User.Upsert(
	User.SlackID.Equals(u.SlackID),
	User.Email.Set(u.Email),
	User.TeamID.Equals(u.TeamID),
	User.SlackID.Set(u.SlackID),
	User.Username.Set(u.Username),
	User.FirstName.Set(u.FirstName),
	User.LastName.Set(u.LastName),
	User.AvatarURL.Set(u.AvatarURL),
	User.Timezone.Set(u.Timezone),
	User.IsTeamOwner.Set(true),
	User.TeamID.Set(tm.ID),
)
```

I agree that this is still more verbose than yours, but I think it's worth to have the type safety here. This is the only reliable way to verify that all required fields are in fact specified. If you change or add a new required field in your data model, you will get warnings or errors for migrations if it would involve a breaking change. I would like to reflect this in the Go client as well, so you are forced to also adapt your code. I bet you that people (myself included) will forget to adapt their code for some .Create or .Upsert calls, Go would default to "" and users will wonder why their data is inconsistent or their queries won't work afterwards.

I think Go is more about simplicity and type safety than verbosity. Of course, it tries to be terse, but not when you have additional type safety – that's why you have verbose error checks because it may be easier to debug later, no generics which forces you to write specific functions for specific data types which means you have extremely high type safety.

Also, another advantage of using methods on each field, is perfect for nullable fields, regardless of queries (.Equals etc.) or setters. For example, if you have a nullable field, we may add `.SetOptional`/`.Opt`/whatever to nullable fields which accepts a string instead of a NullString/pointer if you want to set an optional value to a fixed string so there's no need a need for extra helper methods. 

If we want to most simple method, I think that would be using structs, especially because IDEs can fill _all_ fields with a single click, and you can even omit the field names. This would have the equal amount of type safety as your proposal, but the same drawbacks (not forced to specify required fields).

Reply:
Related: https://github.com/prisma/photongo/issues/8